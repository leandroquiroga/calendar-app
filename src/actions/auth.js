import { fetchNotToken, fetchWithToken } from "../helpers/fetchAPI"
import { types } from "../types/types";

const login = (user) => ({
  type: types.authLogin,
  payload: user
});

const checkingFinish = () => ({ type: types.authChekingFinish });

const logout = () => ({ type: types.authLogout });

const eventLogout = () => ({ type: types.eventLogout });

export const authStartLogin = (email, password, setMsgError) => {
  return async (dispatch) => {
    const response = await fetchNotToken('auth/login', { email, password }, 'POST');
    const body = await response.json();

    // En caso de que haya un error en el login retorna el mensaje de error.
    if (!body.ok) return setMsgError(body.msg);

    localStorage.setItem('token', body.token);
    localStorage.setItem('token-init-date', new Date().getTime());
    dispatch(login({
      uid: body.uid,
      name: body.name
    }));
  }
} 

export const startRegister = (email, password, name, setMsgError) => {
  return async (dispatch) => {
    const response = await fetchNotToken('auth/register', { email, password, name }, 'POST');
    const body = await response.json();

    // En caso de que haya un error en el registro retornara un mensaje de error 
    if (!body.ok) return setMsgError(body.msg);

    localStorage.setItem('token', body.token);
    localStorage.setItem('token-init-date', new Date().getTime());
    dispatch(login({
      uid: body.uid,
      name: body.name,
    }));
  };
}


export const startCheking = () => {
  return async (dispatch) => {
    const response = await fetchWithToken('auth/renew');
    const body = await response.json();

    if (!body.ok) {
      // Cambia el checking a true y retorna un mensaje de error desde el backend
      return dispatch(checkingFinish());
    };

    localStorage.setItem('token', body.token);
    localStorage.setItem('token-init-date', new Date().getTime());

    dispatch(login({
      uid: body.uid,
      name: body.name,
    }));
  };
};


export const startUserLogut = () => {
  return (dispatch) => {
    // Debe eliminar el token
    localStorage.removeItem('token');
    localStorage.removeItem('token-init-date');

    dispatch(eventLogout())
    dispatch(logout());
  }
};