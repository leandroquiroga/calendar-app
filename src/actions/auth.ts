import { Dispatch, SetStateAction } from "react";

import { AppDispatch } from "../store/store";
import { fetchNotToken, fetchWithToken } from "../helpers/fetchAPI"
import { types } from "../types/types";

export interface UserProps {
  email?: string;
  password?: string;
  name: string;
  uid: string;
};

export interface UserLogin {
  email: string;
  password: string;
  setMsgError: Dispatch<SetStateAction<string>>
};

export interface UserRegister {
  email: string;
  password: string;
  name?: string;
  setMsgError: Dispatch<SetStateAction<string>>
};

const login = (user: UserProps) => ({
  type: types.authLogin,
  payload: user
});

const checkingFinish = () => ({ type: types.authChekingFinish });

const logout = () => ({ type: types.authLogout });

const eventLogout = () => ({ type: types.eventLogout });

export const authStartLogin = (
  email: string,
  password: string,
  setMsgError: Dispatch<SetStateAction<string>>
): any => {
  return async (dispatch: AppDispatch) => {
    const response = await fetchNotToken(
      "auth/login",
      { email, password },
      "POST"
    );
    const body = await response.data;

    // En caso de que haya un error en el login retorna el mensaje de error.
    if (!body.ok) return setMsgError(body.msg);

    localStorage.setItem("token", body.token);
    localStorage.setItem("token-init-date", new Date().getTime().toString());
    dispatch(
      login({
        uid: body.uid,
        name: body.name,
      })
    );
  };
}; 

export const startRegister = (
  email: string,
  password: string,
  name: string,
  setMsgError: Dispatch<SetStateAction<string>>,
): any => {
  return async (dispatch: AppDispatch) => {
    const response = await fetchNotToken(
      "auth/register",
      { email, password, name },
      "POST"
    );
    const body = await response.data;

    // En caso de que haya un error en el registro retornara un mensaje de error
    if (!body.ok) return setMsgError(body.msg);

    localStorage.setItem("token", body.token);
    localStorage.setItem("token-init-date", new Date().getTime().toString());
    dispatch(
      login({
        uid: body.uid,
        name: body.name,
      })
    );
  };
};

// Buscar una mejor definicion para el tipo de retorno de la funcion
export const startCheking = (): any => {
  return async (dispatch: AppDispatch) => {
    const response = await fetchWithToken("auth/renew", "GET");
    const body = await response.data;
    if (!body.ok) {
      // Cambia el checking a true y retorna un mensaje de error desde el backend
      return dispatch(checkingFinish());
    }

    localStorage.setItem("token", body.token);
    localStorage.setItem("token-init-date", new Date().getTime().toString());

    dispatch(
      login({
        uid: body.uid,
        name: body.name,
      })
    );
  };
};

// Buscar una mejor definicion para el tipo de retorno de la funcion
export const startUserLogut = ():any => {
  return (dispatch: AppDispatch) => {
    // Debe eliminar el token
    localStorage.removeItem("token");
    localStorage.removeItem("token-init-date");

    dispatch(eventLogout());
    dispatch(logout());
  };
};