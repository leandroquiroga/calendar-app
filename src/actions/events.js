import { types } from './../types/types';
import { fetchWithToken } from '../helpers/fetchAPI';
import { prepareEvents } from '../helpers/prepareEvents';

const eventAddNew = (event) => ({
  type: types.eventAddNew, 
  payload: event,
});

const eventLoaded = (event) => ({
  type: types.eventLoaded,
  payload: event
});

const updatedEvent = (event) => ({
  type: types.eventUpdateNotes,
  payload: event
});

const deleteEventNotes = () => ({ type: types.eventDeleteNotes });

export const eventStartAddNew = (event) => {
  return async (dispatch, getState) => {

    const { uid, name } = getState().auth;
    
    try {
      const response = await fetchWithToken('events', event, 'POST');
      const body = await response.json();

      if (body.ok) {
        event.id = body.event.id
        event.user = {
          uid,
          name,
        };
        dispatch(eventAddNew(event));
      }
      
    } catch (error) {
      console.log(error)
    }
  };
};

export const eventSetActive = (event) => ({
  type: types.eventSetActive,
  payload: event
})

export const clearEventActive = () => ({ type: types.clearActiveEvent });

export const eventStartLoading = () => {
  return async(dispatch) => {
    try {
      const response = await fetchWithToken('events');
      const body = await response.json(); 
      const events = prepareEvents(body.event);
      
      dispatch(eventLoaded(events))
    } catch (error) {
      console.log(error)
    }
  };
};

export const eventStartUpdate = (event) => {
  return async (dispatch) => {
    try {
      const response = await fetchWithToken(`events/${event.id}`, event, 'PUT');
      const body = await response.json();

      if (body.ok)
        return dispatch(updatedEvent(event));

      console.log(body.msg);
    } catch (error) {
      console.log(error)
    }
  }
};

export const eventStartDelete = () => {
  return async (dispatch, getState) => {

    const { id } = getState().calendar.activeEvent;
    
    try {
      const response = await fetchWithToken(`events/${id}`, {}, 'DELETE');
      const body = await response.json();
      
      if (body.ok)
        return dispatch(deleteEventNotes());
      
      console.log(body.msg)
      
    } catch (error) {
      console.log(error)
    }
  }
};