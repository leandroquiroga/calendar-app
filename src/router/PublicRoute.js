import React from 'react'
import { Navigate } from 'react-router-dom';
import { useSelector } from 'react-redux/es/hooks/useSelector';

export const PublicRoute = ({ children }) => {
  const { uid } = useSelector(state => state.auth);

  return (!uid)
    ? children
    : <Navigate to='/' />
};