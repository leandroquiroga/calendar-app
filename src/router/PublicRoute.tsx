import React from 'react';
import { Navigate } from 'react-router-dom';
import { useSelector } from 'react-redux';

export const PublicRoute = ({ children }: any) => {
  const { uid } = useSelector((state: any) => state.auth);

  return !uid ? children : <Navigate to="/" />;
};