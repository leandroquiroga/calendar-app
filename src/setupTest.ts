import '@testing-library/jest-dom';
import Enzyme from 'enzyme';
import ReactAdapter from '@wojtekmaj/enzyme-adapter-react-17';
import { createSerializer,  } from 'enzyme-to-json';

Enzyme.configure({ adapter: new ReactAdapter() });

expect.addSnapshotSerializer(createSerializer({mode: 'deep'}) as any)