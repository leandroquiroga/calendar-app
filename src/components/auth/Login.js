import React, { useEffect, useState } from 'react'
import { Link } from 'react-router-dom';
import { useForm } from './../../hooks/useForm';
import { authStartLogin } from '../../actions/auth';
import { useDispatch } from 'react-redux';
import logoLogin from '../../assets/login.svg';

export const Login = () => {

  const dispatch = useDispatch();
  const [msgError, setMsgError] = useState('');
  const [buttonDisabled, setButtonDisabled] = useState(false);

  const [{ email, password }, handleChange] = useForm({
    email: '',
    password: '',
  });

  const handleSubmit = (e) => {
    e.preventDefault();

    dispatch(authStartLogin(email, password, setMsgError));
  };
  
  useEffect(() => {
    if ([email, password].includes('')) {
      setButtonDisabled(true);
      return
    };
    setButtonDisabled(false)
  }, [email, password, buttonDisabled])

  return (
    <section
      className='row p-0 m-0'
      style={{ height: 100 + 'vh', width: 100 + '%' }}
    >
      <article className='col-md-6 d-flex flex-column justify-content-center align-items-center'>
        <div className='border rounded-circle overflow-hidden shadow-sm p-3'>
          <img
            className='img-fluid'
            style={{ width: 250 + 'px' }}
            src={logoLogin}
            alt='login'
          />
        </div>

        <div className='text-center p-2 mt-3'>
          <p className='fs-5 fw-bold'> !Bienvenido a Calendar! </p>
          <small className='fw-normal fs-6'> Inicie sesión para abrir su agenda web. </small>
          <hr />
          <span className='m-0'>
            ¿No tienes cuenta? Haz click en {''}
            <Link
              to='/register'
              className='text-decoration-none'
            >
              Registrarme
            </Link>
          </span>
        </div>

      </article>
      <article className='col-md-6 bg-primary d-flex flex-column justify-content-center align-items-center'>
        <h1 className='text-white fs-3 p-4'> Inicia Sesión </h1>
        <form
          className='form-control my-5 form_container bg-white border-0 rounded p-3'
          onSubmit={handleSubmit}
        >
          <div className='mb-3 d-flex flex-column'>
            <label
              htmlFor='email'
              className='fw-normal form-label'
            >
              Correo electronico:
            </label>
            <input
              className='border-0 border-bottom form_container__field'
              id='email'
              placeholder='Ingrese su email...'
              type='email'
              name='email'
              value={email}
              onChange={handleChange}
            />
          </div>
          <div className='mb-3 d-flex flex-column'>
            <label
              htmlFor='password'
              className='fw-normal form-label'
            >
              Contraseña:
            </label>
            <input
              className='border-0 border-bottom form_container__field'
              id='password'
              placeholder='Ingrese su contraseña...'
              type='password'
              name='password'
              value={password}
              onChange={handleChange}
            />
          </div>

          <button
            type='submit'
            disabled={buttonDisabled}
            className={
              (buttonDisabled)
                ? 'btn btn-primary rounded p-2 form_container__button form_container__button-disabled'
                : 'btn btn-primary rounded p-2 form_container__button'}
          >
            Iniciar sesión
          </button>
          {
            (msgError !== '') && <p className='text-center mt-2 text-danger fs-6'>{msgError}</p>
          }
        </form>
      </article>
    </section>
  )
};