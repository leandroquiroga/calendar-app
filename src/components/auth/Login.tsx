import React, { useState } from "react";
import { Form, Formik } from 'formik';
import { Link } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { AnyAction, Dispatch } from 'redux';

import { Input } from '../form/Input';
import { Button } from '../form/Button';
import { authStartLogin } from '../../actions/auth';
import logoLogin from '../../assets/login.svg';

export interface InitialValues {
  email: string;
  password: string;
}

export const valueForm: InitialValues = {
  email: '',
  password: '',
};

export const Login = () => {

  const dispatch = useDispatch<Dispatch<AnyAction>>();
  const [msgError, setMsgError] = useState<string>('');


  const handleSubmit = ( values: InitialValues) => {
    const { email, password } = values;
    dispatch(authStartLogin(email, password, setMsgError));
  };

  return (
    <section
      className="row p-0 m-0"
      style={{ height: 100 + "vh", width: 100 + "%" }}
    >
      <article className="col-md-6 d-flex flex-column justify-content-center align-items-center">
        <div className="border rounded-circle overflow-hidden shadow-sm p-3">
          <img
            className="img-fluid"
            style={{ width: 250 + "px" }}
            src={logoLogin}
            alt="login"
          />
        </div>

        <div className="text-center p-2 mt-3">
          <p className="fs-5 fw-bold"> Bienvenido a Calendar </p>
          <small className="fw-normal fs-6">
            {" "}
            Inicie sesión para abrir su agenda web.{" "}
          </small>
          <hr />
          <span className="m-0">
            ¿No tienes cuenta? Haz click en {""}
            <Link to="/register" className="text-decoration-none">
              Registrarme
            </Link>
          </span>
        </div>
      </article>
      <article className="col-md-6 bg-primary d-flex flex-column justify-content-center align-items-center">
        <h1 className="text-white fs-3 p-4"> Inicia Sesión </h1>

        <Formik
          initialValues={valueForm}
          onSubmit={(values) => handleSubmit(values)}
        >
          {(formik) => (
            <Form
              className="form-control my-5 form_container bg-white border-0 rounded p-3"
              noValidate
            >
              <Input
                label="Email"
                name="email"
                type="email"
                styles="border-0 border-bottom form_container__field"
                placeholder="Ingrese su correo electronico"
              />
              <Input
                label="Contraseña"
                name="password"
                type="password"
                styles="border-0 border-bottom form_container__field"
                placeholder="Ingrese su contraseña"
              />

              <Button
                type="submit"
                styles="btn btn-primary rounded p-2 form_container__button"
                value="Iniciar sesión"
              />
              {msgError !== "" && (
                <p className="text-center mt-2 text-danger fs-6">{msgError}</p>
              )}
            </Form>
          )}
        </Formik>
      </article>
    </section>
  );
};