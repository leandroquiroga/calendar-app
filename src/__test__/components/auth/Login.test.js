import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { Login } from '../../../components/auth/Login';
import { PublicRoute } from '../../../router/PublicRoute';
import { authStartLogin } from '../../../actions/auth';

import '@testing-library/jest-dom';
import '../../../setupTest';

jest.mock('../../../actions/auth', () => ({
  authStartLogin: jest.fn()
}));
const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
  auth: {
    uid: null
  }
};
let store = mockStore(initState);

store.dispatch = jest.fn();

const wapper = mount(
  <Provider store={store}>
    <MemoryRouter initialEntries={['/login']}>
    <Routes>
      <Route path='login' element={
        <PublicRoute>
          <Login />
        </PublicRoute>
        }/>
    </Routes>
    </MemoryRouter>
  </Provider>
);


describe('Test on <Login />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  test('this test should make a snapshot with component', () => {
    expect(wapper).toMatchSnapshot();
  });

  test('this test should show atributte disabled of the button in true', () => {
    const btn = wapper.find('button');
    expect(btn.props().disabled).toBeTruthy();
  });

  test('this test should run the dispatch of authStartLogin', async () => {


    wapper.find('input[name="email"]').simulate('change', {
      target: {
        name: 'email',
        value: 'tester@testing.com'
      }
    });

    wapper.find('input[name="password"]').simulate('change', {
      target: {
        name: 'password',
        value: '123456'
      }
    });

    const btn = wapper.find('button');
    expect(btn.props().disabled).not.toBeTruthy();

    wapper.find('form').prop('onSubmit')({
      preventDefault() { }
    });

    expect(authStartLogin).toHaveBeenCalled();
  });
});