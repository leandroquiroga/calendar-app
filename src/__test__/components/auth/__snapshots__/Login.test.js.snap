// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <Login /> this test should make a snapshot with component 1`] = `
<section
  className="row p-0 m-0"
  style={
    Object {
      "height": "100vh",
      "width": "100%",
    }
  }
>
  <article
    className="col-md-6 d-flex flex-column justify-content-center align-items-center"
  >
    <div
      className="border rounded-circle overflow-hidden shadow-sm p-3"
    >
      <img
        alt="login"
        className="img-fluid"
        src="login.svg"
        style={
          Object {
            "width": "250px",
          }
        }
      />
    </div>
    <div
      className="text-center p-2 mt-3"
    >
      <p
        className="fs-5 fw-bold"
      >
         !Bienvenido a Calendar! 
      </p>
      <small
        className="fw-normal fs-6"
      >
         Inicie sesión para abrir su agenda web. 
      </small>
      <hr />
      <span
        className="m-0"
      >
        ¿No tienes cuenta? Haz click en 
        <a
          className="text-decoration-none"
          href="/register"
          onClick={[Function]}
        >
          Registrarme
        </a>
      </span>
    </div>
  </article>
  <article
    className="col-md-6 bg-primary d-flex flex-column justify-content-center align-items-center"
  >
    <h1
      className="text-white fs-3 p-4"
    >
       Inicia Sesión 
    </h1>
    <form
      className="form-control my-5 form_container bg-white border-0 rounded p-3"
      onSubmit={[Function]}
    >
      <div
        className="mb-3 d-flex flex-column"
      >
        <label
          className="fw-normal form-label"
          htmlFor="email"
        >
          Correo electronico:
        </label>
        <input
          className="border-0 border-bottom form_container__field"
          id="email"
          name="email"
          onChange={[Function]}
          placeholder="Ingrese su email..."
          type="email"
          value=""
        />
      </div>
      <div
        className="mb-3 d-flex flex-column"
      >
        <label
          className="fw-normal form-label"
          htmlFor="password"
        >
          Contraseña:
        </label>
        <input
          className="border-0 border-bottom form_container__field"
          id="password"
          name="password"
          onChange={[Function]}
          placeholder="Ingrese su contraseña..."
          type="password"
          value=""
        />
      </div>
      <button
        className="btn btn-primary rounded p-2 form_container__button form_container__button-disabled"
        disabled={true}
        type="submit"
      >
        Iniciar sesión
      </button>
    </form>
  </article>
</section>
`;
